### ISSUE TITLE
[DEPR] Devstack

### ISSUE THREAD

**Issue by kdmccormick** [openedx/public-engineering#247]
## Timeline

* Proposed: 2022-03-01

* Updated Acceptance Date: 2024-03-12

### February 2024 Update
* We are planning on archiving the devstack repo and moving the repository into 'unsupported' once this deprecation is complete.
* Anyone wishing to continue to use devstack will have to make their own forks in order to make changes or keep it updated/maintained.
* 2U will likely maintain its own fork, but there will be a number of 2U-specific changes and will not be open to outside development. It is possible that this fork will be made private.

## Replacement

The replacement for Devstack is Tutor. [Tutor](https://github.com/overhangio/tutor) is "the Docker-based Open edX distribution designed for peace of mind". It is both a deployment tool and a development environment. Tutor has been the only community-support production deployment method since Maple, although the Open edX docs and [website](https://openedx.org/community/develop/) do not officially recommend it as a development environment yet.

### Rationale

Devstack has several issues:

* It is not pluggable. This leads to two issues:
  * It is large. Out of the box, it comes with a dozen IDAs and another dozen MFEs; new Open edX developers generally needs two IDAs (LMS and CMS) a few MFEs. In the past couple years, improvements have been made such that only a subset of IDAs are run by default. Still, cloning and provisioning are bound by the full weight of all services.
  * Some IDAs like [enterprise-catalog](https://github.com/openedx/enterprise-catalog), in order to avoid being part of the large Devstack core, make a plugin-esque docker-compose.yml file that hooks into Devstack. This complicates the developer experience on those IDAs and also makes it harder to modify Devstack itself (because of the implicit plugin interface that such IDAs are using).
* Its images are slow to download. The Docker images used by devstack are built by running Ansible in a Dockerfile instead of using native Dockerfile commands. This means that the images are large and make poor use of the Docker layer cache, and thus take a long time for Devstack users to download.
* Many users find it difficult to diagnose and fix Devstack errors. Often, they find it more time/effort-efficient to destroy their volumes and provision from scratch instead of fixing the immediate issue.
* Provisioning takes a long time. This exacerbates the previous issue.
* There are several "upkeep" steps that must be independently done on a regular basis in order to keep Devstack working reliably. These include git-pulling devstack, git-pulling multiple app repos, pulling Docker images, and running migrations.
* Setting up a service in Devstack is completely tangential to setting it up for production deployment. Once it is set up, it takes attention and effort to keep the setup consistent between Devstack and production.

Tutor shows promise in many of these areas:
* It features a documented plugin API. Many IDAs (forums, discovery, credentials, et al) are implemented as plugins to Tutor. The core of Tutor includes just LMS and CMS, and [there is even work being discussed to factor those out into plugins](https://github.com/overhangio/2u-tutor-adoption/issues/10).
* Tutor's Dockerfiles are written using native Dockerfile syntax and are intentionally designed to be as small and cache-friendly as reasonably possible.
* Tutor has [actively-maintained documentation](https://docs.tutor.overhang.io) including a Troubleshooting section. Beyond that, Tutor has a dedicated forum community where maintainers and users help each other. Tutor also makes a point to have a consistent, documented, and debuggable CLI. For example, the `tutor dev` segment of the CLI is implemented entirely on top of `docker-compose` and transparently prints out the underlying `docker-compose` commands that it runs to aid in user debugging, using terminal colors to call out commands vs tutor logging vs actual output.
* Once Tutor is installed, provisioning involves one command (`tutor local quickstart`) and runs fairly quickly. Provisioning can be safely re-run without destroying existing data.
* Because Tutor uses the code in the Docker images by default, the user only needs to regularly git-pull the Tutor repo and the repos that they host-mount in, if any.
* Since Tutor is also the community-supported deployment method, one can support development and and production usage of an IDA all in a single plugin. Furthermore, using Tutor for both dev and prod allows operators to keep their dev and prod environments very similar with less effort.

For areas where Tutor could be improved, there is an [ongoing Tutor Adoption Initiative](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3315335223/Tutor+Adoption+Initiative) which aims to support the transition from Devstack to Tutor through a mix of education, plugin development, and core Tutor improvements. Within the [initiative roadmap](https://github.com/orgs/openedx/projects/10), the "Dev Quality-of-Life" epic contains stories that should ideally close any existing gaps between Tutor and Devstack.

### Migration

Many developers can start using Tutor now, and many already are). Particularly, developers whose projects are encompassed by the LMS, CMS, MFEs, and [IDAs covered by existing official plugins](https://overhang.io/tutor/plugins) can generally transition their workflows over to Tutor. On the other hand, for teams that run IDAs with less community adoption (notably, teams at 2U), several new Tutor plugins will be needed in order for Tutor to fully replace Devstack. 

There are no plans to help users migrate data from a Devstack instance to a Tutor instance. We recommend that folks start fresh when switching from Devstack to Tutor.

### Deprecation steps

* Update [the development page on open.edx.org](https://openedx.org/community/develop/) to recommend Tutor instead of Devstack.
* `Ctrl+F` the docs for "Devstack", and where appropriate, update them to point at Tutor.
* Add a deprecation warning to [devstack's readme](github.com/openedx/devstack)
* Make devstack's `dev.up` Makefile target print a deprecation warning
* Update [the configuration docker role's readme](https://github.com/openedx/configuration/blob/master/docker/README.rst) with a deprecation message.
* In the description of https://dockerhub.com/u/edxops, warn that all images in the org are deprecated.

### Removal steps

* Archive https://github.com/openedx/devstack 
* Delete the various `devstack.py` and `development.py` settings files unless otherwise depended upon.
* Delete the various `docker-compose.yml` files and Makefile targets that some IDAs (eg enterprise-catalog) use in order to integrate with Devstack.
* Delete the `docker/` role (incl. Dockerfiles) from github.com/openedx/configuration.
* 2U-specific stuff:
  * Disable Jenkins jobs that build `edxops/*` Docker images and delete the code that defines said jobs.
  * Decommission the `edxops` DockerHub organization?

**Comment by kdmccormick** [comment]
Re-communication happened here: https://discuss.openedx.org/t/deprecation-removal-devstack-update/12384

Aaaand it's archived: https://github.com/openedx-unsupported/devstack

I moved this ticket to public-engineering, since we have a few more clean-up steps before we can close it out.

**Comment by kdmccormick** [comment]
With the devstack repository archived, the obvious next question is: **_When will devstack.py, env.devstack, and any other devstack-specific files be removed from upstream repositories?_** I realize that removing those files will any forks of devstack, unless the forks adapt by loading the files from somewhere new.

I propose that the files should remain in-place until the Sumac branch cut on **October 9th**, after which they are fair game for removal. Happy to discuss different dates, a phased removal, or something else. Let me know what you think @dianakhuang @regisb @cmltaWt0 @felipemontoya @robrap @feanil .

**Comment by feanil** [comment]
October 9th seems like plenty far int the future, but I think we should put deprecation warnings in those files now with the agreed upon dates so people who touch them have some chance of knowing that they're gonna disappear.

**Comment by robrap** [comment]
Thanks @kdmccormick. I'm hopeful we can work that into our plan at this point, but tagging @jristau1984 and @spencertiberi, who need to be involved in any deadline agreements. Thanks.

**Comment by kdmccormick** [comment]
I'd like to settle on a timeline for removal of these auxillary files. If there any objections or alternative suggestions to Oct 9 for the "OK to start removing date", please raise them by the end of next week (Apr 5).

**Comment by jristau1984** [comment]
At this time, I have no concerns with having the removal work begin after the Sumac cut, currently scheduled on Oct 9.

Can this date be added to the timeline section of the issue above?  Thanks!

**Comment by robrap** [comment]
@jristau1984: Just to ensure you are aware, this is not just busy work, but requires designing a plan for how we can accomplish this.

**Comment by kdmccormick** [comment]
> Can this date be added to the timeline section of the issue above? Thanks!

Done.

When I get a chance, I'll start a draft PR of these changes for edx-platform, so that devstack users can get a sense of which files will be going away.

**Comment by kdmccormick** [comment]
Deprecation warnings for devstack settings files in edx-platform: https://github.com/openedx/edx-platform/pull/34795

**Comment by iamsobanjaved** [comment]
Hi @kdmccormick 

I’m working on the 2U side to move Devstack-related settings out of the Open edX repositories and into a different location within the edX GitHub organization. While doing this I was wondering how Tutor has done the development environment setup, I noticed that including the LMS, CMS, and other plugins like discovery and credentials, relies heavily on the devstack.py or development.py modules. These services use devstack.py as a base and then modify a few settings variables.

Is there currently any effort underway for the Tutor to stop depending on these files? Or, as mentioned in the Removal Steps, should we refrain from deleting them if they are still in use?

> Delete the various devstack.py and development.py settings files unless otherwise depended upon.

Tutor using devstack files as a base:
- https://github.com/overhangio/tutor/blob/master/tutor/templates/apps/openedx/settings/lms/development.py#L3
- https://github.com/overhangio/tutor/blob/master/tutor/templates/apps/openedx/settings/cms/development.py#L3
- https://github.com/overhangio/tutor-discovery/blob/master/tutordiscovery/templates/discovery/apps/settings/tutor/development.py#L1
- https://github.com/overhangio/tutor-credentials/blob/master/tutorcredentials/templates/credentials/apps/credentials/settings/development.py#L1

**Comment by regisb** [comment]
I hadn't thought about the fact that getting rid of devstack meant removing devstack.py. Do we all agree that we still need platform-agnostic development settings, though? I would expect that devstack.py would be replaced by development.py settings files that provide good defaults for development. If we do, then we can configure tutor to point to these new development modules instead of devstack.py.

**Comment by kdmccormick** [comment]
Clarification: `development.py` would be a suitable base for Devstack settings, but it would not work out-of-the-box for Devstack because it would not include the specific ports and hostnames that Devstack needs, and it may not have the exact same toggles that Devstack users expect. So, Devstack would still need a custom settings file for LMS and CMS. In the meantime, I recommend making sure that Devstack has the capability to define a custom DJANGO_SETTINGS_MODULE.

**Comment by iamsobanjaved** [comment]
Yeah that's what exactly I wanted, a suitable base for dev settings that can be further extended. Also, we are planning to use YAML to override devstack-related settings and It would be nice to have a development.py that can be extended/overridden by a YAML file.

**Comment by kdmccormick** [comment]
Great. In order to keep development.py straightforward, I am inclined _not_ to load YAML into it, but I can provide a YAML snippet that you could add to Devstack's custom settings module in order to retain YAML support.

**Comment by regisb** [comment]
Thank you! I'm looking forward to a future with streamlined development settings. 

On loading YAML files: I'm also not too inclined on doing that...  but I'm sure you'll be able to do that within devstack.py.

It's awesome that we can now switch to local.openedx.io! As we [learned previously](https://github.com/overhangio/tutor/pull/955/files), the switch is actually very easy. I'll add that to our roadmap.

**Comment by dianakhuang** [comment]
@UsamaSadiq here's where we're discussing the devstack.py file conversion to development.py. We will need to build out the ability to load a custom settings file for devstack and then use that to load in the devstack-specific setings. We will probably need to change our plans accordingly.

**Comment by kdmccormick** [comment]
I haven't had time to make the `(lms,cms)/envs/development.py` files yet. As I mentioned before, we'll make sure to give a 1-month overlap time where both devstack.py and development.py are available, before removing devstack.py. I've updated the ticket description, and I'll post an update here when both development.py files are ready for review.

**Comment by robrap** [comment]
[inform] 
- The cookie-cutter ultimately needs to be updated to match whatever final pattern you want around docker for new IDAs.
- Related, when creating the new codejail-service, we simply removed all the docker/devstack-related code. It may one day want to get updated with whatever pattern is created.

**Comment by kdmccormick** [comment]
@robrap 

* Good point, I added a task to the issue description
* Thank you for doing that!

**Comment by kdmccormick** [comment]
Update: While devstack itself is deprecated, there are still many tendrils to clean up in the openedx org , such as:
* outdated docs.
* devstack django settings files , whose removal is blocked by https://github.com/openedx/edx-platform/issues/36215
*  [devstack-specific env.development files](https://github.com/openedx/frontend-app-authoring/blob/master/.env.development) in the frontend-apps. Their removal is not imminent, but it also not blocked by any upstream concerns. I encourage any users of devstack forks to consider how they could either make those files tooling-agnostic (perhaps using local.openedx.io) or how they could operate devstack without those files being in the upstream repos.

**Comment by robrap** [comment]
@kdmccormick: I'm wondering if we could do away with the "blocked" status, and these tickets are helping me get more clear.

I think this ticket could move to Transition Unblocked and when updating to the new DEPR template, could get a task list for unblocking removal, and a task list for completing the removal. The first list would be required to get to Breaking Changes Unblocked. The Review After Date could simply be used as some future check-in date. Thoughts?

**Comment by kdmccormick** [comment]
Yes, I think that sounds reasonable! I updated the ticket status. I can add that info to the the tasklist in the future (I plan to update all my tickets in one go when I have a couple spare hours).

**Comment by dianakhuang** [comment]
Docker compose files still kicking around: https://github.com/search?type=code&q=(org%3Aopenedx)+NOT+is%3Aarchived+path%3Adocker-compose.yml

**Comment by dianakhuang** [comment]
This is waiting for new development environment settings files to be created. Blocked on the edx-platform settings refactoring.
